superset/charts/data/query_context_validator.py

```diff
@@ -0,0 +1,168 @@
+# Licensed to the Apache Software Foundation (ASF) under one
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+# pylint: disable=invalid-name, no-self-use, too-few-public-methods
+from __future__ import annotations
+
+from typing import List, Optional, Set, TYPE_CHECKING
+
+from superset.charts.data.commands.get_data_command import QueryContextValidator
+from superset.errors import ErrorLevel, SupersetError, SupersetErrorType
+from superset.exceptions import SupersetSecurityException
+from superset.sql_parse import ParsedQuery, Table
+
+if TYPE_CHECKING:
+    from superset import SupersetSecurityManager
+    from superset.common.query_context import QueryContext
+    from superset.common.query_object import QueryObject
+    from superset.connectors.base.models import BaseDatasource
+    from superset.connectors.sqla.models import SqlaTable
+    from superset.datasets.dao import DatasetDAO
+    from superset.models.core import Database
+    from superset.typing import Metric
+
+
+class QueryContextValidatorImpl(QueryContextValidator):
+    _dataset_dao: DatasetDAO
+    _security_manager: SupersetSecurityManager
+
+    def __init__(
+        self, dataset_dao: DatasetDAO, security_manager: SupersetSecurityManager
+    ):
+        self._dataset_dao = dataset_dao
+        self._security_manager = security_manager
+
+    def validate(self, query_context: QueryContext) -> None:
+        self._validate_actor_can_access(query_context)
+        self._validate_queries_context(query_context)
+
+    def _validate_actor_can_access(self, query_context: QueryContext) -> None:
+        sql_database: Optional[Database] = query_context.get_database()
+        if sql_database is not None:
+            self._validate_when_context_based_on_sql_database(
+                query_context, sql_database
+            )
+        else:
+            self._security_manager.raise_for_access(query_context=query_context)
+
+    def _validate_when_context_based_on_sql_database(
```

Comment:
This has come up in previous review comments from me. While I agree that it's good practice to break out logical chunks to improve readability and enforce separation of concerns, there is a point when abstraction can actually decrease readability. Each nested call causes additional cognitive overhead for the reviewer, and dilutes the relevance of other more important functions/methods. As a rule of thumb, if a private method is 1) only used once 2) consists of less than 5 lines of code 3) further branches out into another private method which also satisfies 1) and 2), I would consider just inlining the code and adding a comment describing what is happening (if needed).

Reply:
Yep, I agree. Too much delegation makes the logic flow hard to follow